metrics/time/launch_times.sh

```diff
@@ -0,0 +1,241 @@
+
+	if [ -n "$CALCULATE_KERNEL" ]; then
+		# Grab the last kernel dmesg time
+		# In our case, we need to find the last real kernel line before
+		# the systemd lines begin. The last:
+		# 'Freeing unused kernel memory' line is a reasonable
+		# 'last in kernel line' to look for.
+		# We make a presumption here that as we are in a cold-boot VM
+		# kernel, the first dmesg is at '0 seconds', so the timestamp
+		# of that last line is the length of time in the kernel.
+		kernel_last_line=$( (fgrep "Freeing unused kernel memory" <<- EOF
+				${workload_result[@]}
+			EOF
+			) | tail -1 )
+
+		if [ -z "$kernel_last_line" ]; then
+			echo "No kernel last line"
+			for l in "${workload_result[@]}"; do
+				echo ">: [$l]"
+			done
+			die "No kernel last line"
+		fi
+
+
+		kernel_period=$(echo $kernel_last_line | awk '{print $2}' | tr -d "]")
+
+		# And we can then work out how much time it took to get to the kernel
+		to_kernel_period=$(bc <<<"scale=3; $(ns_to_s $workload_period) - $kernel_period")
+	fi
+
+	# And store the results...
+	local json="$(cat << EOF
+	{
+		"total": {
+			"Result": $(ns_to_s $total_period),
+			"Units" : "s"
+		},
+		"to-workload": {
+			"Result": $(ns_to_s $workload_period),
+			"Units" : "s"
+		},
+		"in-kernel": {
+			"Result": $kernel_period,
+			"Units" : "s"
+		},
+		"to-kernel": {
+			"Result": $to_kernel_period,
+			"Units" : "s"
+		},
+		"to-quit": {
+			"Result": $(ns_to_s $shutdown_period),
+			"Units" : "s"
+		}
+	}
+EOF
+)"
+	metrics_json_add_array_element "$json"
+
+	# If we are doing an (optional) scaling test, then we launch a permanent container
+	# between each of our 'test' containers. The aim being to see if our launch times
+	# are linear with the number of running containers or not
+	if [ -n "$SCALING" ]; then
+		docker run -d ${IMAGE} sh -c "tail -f /dev/null"
+	fi
+}
+
+init () {
+	TEST_ARGS="image=${IMAGE} runtime=${RUNTIME} units=seconds"
+
+	# We set the generic name here, but we save the different time results separately,
+	# and append the actual detail to the name at the time of saving...
+	TEST_NAME="boot times"
+
+	# If we are scaling, note that in the name
+	[ -n "$SCALING" ] && TEST_NAME="${TEST_NAME} scaling"
+
+	[ -n "$NONETWORKING" ] && NETWORK_OPTION="--network none" && \
+		TEST_NAME="${TEST_NAME} nonet"
+
+	echo "Executing test: ${TEST_NAME} ${TEST_ARGS}"
+	check_cmds "${REQUIRED_CMDS[@]}"
+
+	if [ "$RUNTIME" == "cor" ] || [ "$RUNTIME" == "cc-runtime" ] || [ "$RUNTIME" == "kata-runtime" ] ; then
+		CALCULATE_KERNEL=1
+		DMESGCMD="; dmesg"
+	else
+		# For non-VM runtimes, we don't use the output of dmesg, and
+		# we have seen it cause some test instabilities, so do not invoke
+		# it if not needed.
+		DMESGCMD=""
+	fi
+
+	# Start from a fairly clean environment
+	init_env
+	check_images "$IMAGE"
+}
+
+help() {
+	usage=$(cat << EOF
+Usage: $0 [-h] [options]
+   Description:
+        This script takes time measurements for different
+	stages of a boot/run/rm cycle
```

Comment:
I see that in the other lines of this description, spaces are used instead of tab